src/persist-client/src/internal/compact.rs

```diff
@@ -468,34 +468,76 @@ where
 
-        // populate our heap with the updates from the first part of each run
-        for (index, (part_desc, parts)) in runs.iter_mut().enumerate() {
+        // populate our heap with the updates from the first part of each run.
+        // we have the memory to fetch each part in parallel, so we spawn a
+        // task to fetch each one.
+        let mut pending_initial_parts = VecDeque::with_capacity(usize::min(
+            runs.len(),
+            cfg.compaction_reads_max_outstanding_parts,
+        ));
+        let mut fetched_initial_parts = Vec::with_capacity(runs.len());
+        for (runs_index, (part_desc, parts)) in runs.iter_mut().enumerate() {
             if let Some(part) = parts.next() {
-                let start = Instant::now();
-                let mut part = fetch_batch_part(
-                    shard_id,
-                    blob.as_ref(),
-                    &metrics,
-                    &metrics.read.compaction,
-                    &part.key,
-                    part_desc,
-                )
-                .await?;
-                timings.part_fetching += start.elapsed();
-                let start = Instant::now();
-                while let Some((k, v, mut t, d)) = part.next() {
-                    t.advance_by(desc.since().borrow());
-                    let d = D::decode(d);
-                    let k = k.to_vec();
-                    let v = v.to_vec();
-                    // default heap ordering is descending
-                    sorted_updates.push(Reverse((((k, v), t, d), index)));
-                    remaining_updates_by_run[index] += 1;
+                let shard_id = shard_id.clone();
+                let blob = Arc::clone(&blob);
+                let metrics = Arc::clone(&metrics);
+                let compaction_metrics = metrics.read.compaction.clone();
+                let key = part.key.clone();
+                let part_desc = (**part_desc).clone();
+
+                pending_initial_parts.push_back((
+                    runs_index,
+                    mz_ore::task::spawn(|| "compaction::initial_fetch_batch_part", async move {
+                        fetch_batch_part(
+                            &shard_id,
+                            blob.as_ref(),
+                            &metrics,
+                            &compaction_metrics,
+                            &key,
+                            &part_desc,
+                        )
+                        .await
+                    }),
+                ));
+
+                if pending_initial_parts.len() > cfg.compaction_reads_max_outstanding_parts {
+                    let start = Instant::now();
+                    let (index, part) = pending_initial_parts
+                        .pop_front()
+                        .expect("pop failed when len was just > some usize");
+                    fetched_initial_parts.push((index, part.await??));
+                    timings.part_fetching += start.elapsed();
```

Comment:
now we no longer sum up all "fetch time" but just time we spend waiting. Would an alternative (but equal) way of measuring this be to create `start` right before the above loop and measure `elapsed` right after the next loop?

Reply:
Yeah that makes so much sense 🤦

Reply:
I know this to well! I often take the existing structure and massage it into what I need, without taking a tiny step back... :sweat_smile: